testng-core/src/main/java/org/testng/internal/thread/TestNGThreadFactory.java

```diff
@@ -1,19 +1,29 @@
 package org.testng.internal.thread;
 
+import java.util.Collection;
+import java.util.WeakHashMap;
 import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
 
 public class TestNGThreadFactory implements ThreadFactory {
 
   private final AtomicInteger threadNumber = new AtomicInteger(1);
   private final String name;
+  private final WeakHashMap<Thread, Object> threads = new WeakHashMap<>();
```

Comment:
Prefer to have a weak hashset like in https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Collections.html#newSetFromMap(java.util.Map)